### PR TITLE
[HOLD] Set option on socket to be lingering

### PR DESCRIPTION
**Pull request by iwiznia**
See https://github.com/Expensify/Web-Expensify/pull/23785 for more context.

I am really unsure if this change is good or we need to pass a flag to only set the option on certain cases.

### CODE REVIEW

File: libstuff/libstuff.cpp [hunk 1]
```diff
@@ -1609,6 +1609,14 @@ int S_socket(const string& host, bool isTCP, bool isPort, bool isBlocking) {
                 STHROW("couldn't set non-blocking");
         }
 
+        // Set the lingering option to enabled and no timeout, this makes it so that once the servers goes offline,
+        // clients that send data to it, get an immediate error instead of getting it later.
+        struct linger so_linger;
+        so_linger.l_onoff = 1;
+        so_linger.l_linger = 0;
+        if (setsockopt(s, SOL_SOCKET, SO_LINGER, &so_linger, sizeof(so_linger)))
```
Comment (flodnv) [inline libstuff/libstuff.cpp]: use a `{` please 😬

**Comment by tylerkaraszewski** [comment]
Can someone explain what this is trying to accomplish? The comment in the code is not very clear to me.

**Comment by quinthar** [comment]
When we do a graceful shutdown of a Bedrock node, Bedrock naturally shuts down the socket from the server side.  However, PHP isn't made immediately aware of the shutdown, meaning it still thinks it's alive -- and due to sticky sockets, will attempt to send a new request on the shutdown socket.  That would be fine if the send clearly failed, but it doesn't: PHP reports the send as having succeeded.  And if it's a non-idempotent command, it cannot be retried.  

**tl;dr shutting down a Bedrock node, even gracefully, results in random command failures, which can result in failed jobs**

**Comment by quinthar** [comment]
This PR disables the `LINGER` feature when closing sockets, meaning that upon graceful shutdown, it will just "abrubtly" close the socket -- and do so in a way that notifies PHP that the socket is in fact closed.  This "works", but I personally don't think we should do this because I fear it'll create even worse problems: the LINGER feature is there to ensure that anything sent on the socket is successfully sent *before* the socket is closed.  So I fear this PR will result in us just randomly dropping responses on graceful shutdown -- which will manifest in *exactly the same way* as the original problem (eg, randomly failed requests, which means randomly failed jobs).

**Comment by quinthar** [comment]
Ideally, I think we should just do a `shutdown(RDWR)` on the socket before closing it.  And I *thought* that would notify the remote side that the socket was closed, such that when when PHP tried to send on it, it would fail.  But that didn't seem to work.

**Comment by quinthar** [comment]
Crazy idea, what if we just send an unprompted response to PHP upon graceful shutdown of Bedrock (eg, a response that is not tied to a request).  So it'd work like this:

1. PHP opens a socket to Bedrock
2. PHP sends `200 OK`
3. Bedrock sends `BlahResponse`
4. Bedrock begins graceful shutdown
5. Bedrock sends `500 Hard close`, and then `shutdown(RDWR)`
6. PHP sends `BlahRequest` at exactly the same time, not realizing that the socket has been closed
7. Rather than receiving the expected `200 OK`, it gets `500 Hard Close`

This way PHP will know that the `BlahRequest` wasn't received, and thus can be safely retried.

**Comment by iwiznia** [comment]
> * Bedrock sends `500 Hard close`, and then `shutdown(RDWR)`

This `500 Hard close` is a custom message we will send?

Anyway, I am not sure that will work either, from my tests both the first send and the first recv from the client return no error even if the server is down, I am unsure we would get the `500 Hard close` in that case, but I will try it out and report back.

**Comment by quinthar** [comment]
Well, we've never had any problems sending responses, so why wouldn't PHP get it?  Basically, the `send()` will seem to work (even though the server is going to ignore it), and then the `recv()` will return `"500 Hard close"`

**Comment by tylerkaraszewski** [comment]
Ok, I've looked into this a bit and come up with another idea.

Bedrock already sends:

```
Connection: close
```

When it's going to close a socket after sending the response, see:
```
        // If we're shutting down, tell the caller to close the connection.
        if (_shutdownState.load() != RUNNING) {
            command.response["Connection"] = "close";
        }
```

However, the bedrock client doesn't look at this. What if we made the bedrock PHP client jsut pay attention to the `Connection: close` header it already receives?

**Comment by iwiznia** [comment]
hmmm maybe? When would we get that connection close though? The scenario is:
- PHP connects to bedrock.
- PHP sends some requests and receives some responses.
- Bedrock shuts down.
- When PHP sends the next request, the server is gone and we receive no response or error from the send or recv.

So, I am not sure php would get that `connection:close`...

**Comment by tylerkaraszewski** [comment]
How long do these sockets actually live, currently, is the lifespan of a socket limited to one web request (which makes multiple auth requests), or is it longer than that?

**Comment by iwiznia** [comment]
Yes, it lasts for as long as the request lives, so it varies (web requests are usually short, jobs might be longer).

**Comment by iwiznia** [comment]
Oh, that makes sense. We can try that out. Will work on it.

**Comment by quinthar** [comment]
Well we should 100% use the `connection: close` header if it's there, agreed.  That might in fact solve everything.
